### Assignment 12 KNN 
Implement a KNN model to classify the animals in to categorie

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold,train_test_split
from sklearn.model_selection import cross_val_score

In [2]:
from sklearn.neighbors import KNeighborsClassifier

In [3]:
zoo_data=pd.read_csv("zoo.csv")
zoo_data.head()

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0    aardvark     1         0     0     1         0        0         1   
1    antelope     1         0     0     1         0        0         0   
2        bass     0         0     1     0         0        1         1   
3        bear     1         0     0     1         0        0         1   
4        boar     1         0     0     1         0        0         1   

   toothed  backbone  breathes  venomous  fins  legs  tail  domestic  catsize  \
0        1         1         1         0     0     4     0         0        1   
1        1         1         1         0     0     4     1         0        1   
2        1         1         0         0     1     0     1         0        0   
3        1         1         1         0     0     4     0         0        1   
4        1         1         1         0     0     4     1         0        1   

   type  
0     1  
1     1  
2     4  
3     1  
4     1

#### Checking the shape

In [4]:
zoo_data.shape

(101, 18)

In [5]:
x=zoo_data.iloc[:,1:17]
y=zoo_data.iloc[:,17]

In [6]:
kfold=KFold(n_splits=10)
model_1=KNeighborsClassifier(n_neighbors=1)
result=cross_val_score(model_1,x,y,cv=kfold)

In [7]:
result.mean()

0.9700000000000001

In [8]:
x_train,y_train,x_test,y_test=train_test_split(x,y,test_size=0.3,random_state=8)

In [9]:
model1=KNeighborsClassifier(n_neighbors=1)
model1

KNeighborsClassifier(n_neighbors=1)

### using gridsearchCV

In [10]:
from sklearn.model_selection import GridSearchCV

In [11]:
kvalues=np.array(range(1,50))
kvalues

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])

In [12]:
temp_dict={"n_neighbors":[kvalues]}
parameters=dict(n_neighbors=kvalues)
parameters
temp_dict

{'n_neighbors': [array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
         18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
         35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])]}

In [13]:
model_2=KNeighborsClassifier()
grid=GridSearchCV(estimator=model_2,param_grid=parameters)

In [14]:
grid.fit(x,y)

C:\Users\Lenovo\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:670: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])})

In [15]:
((grid.best_params_) , (grid.best_score_))

({'n_neighbors': 1}, 0.96)

### Conclusion -

#### Our Accuracy is 96 % when we take 1 neighboring classifier